add option to select sections with multiselect

florisie opened the pull request:
Currently it is only possible to configure templates for your sections (in the "kubernetes-overview-custom-views-alist" a-list) to be displayed on the overview page and subsequently select one of your templates (using lowercase "v").

However, I thought it would also be nice to simply select several resources more dynamically (using a multiselect instead of a "single select" menu). So that you can simply select, for example, pods, configmaps and secrets from the multiselect menu to show on the overview page.

With this PR this option is added. It is bound to the "V" key (capital v) (instead of lowercase v that is used for the "template select option).

jinnovation reviewed (changes requested):
Thanks for this idea.

I recommend reframing this PR in terms of the existing "views" paradigm -- or "template," as you call it. In other words: instead of selecting multiple resources to display at once, select multiple *views* to display at once (keeping in mind that these views can, as is currently the case, line up one-to-one with resources). 

This would remove the need to manually maintain a static list of resources (as you've done here w/ `kubernetes-overview-resource-list`), an existing anti-pattern that we're actively trying to move away from w/ tasks like #69 and #235. I imagine this would also allow us to expand the scope of the existing `kubernetes-overview-set-sections` function rather than implement another keybinding that *basically* does the same thing.

jinnovation commented:
Some broader notes, like I mentioned as well in your other PR #277:
- Add unit test(s);
- Add documentation in `docs/` on how to use this functionality. I recommend adding to `docs/how-to.md`;
- Add a corresponding entry to `docs/CHANGELOG.md`.

Once again, thanks for contributing this. Excited to see increased contributor activity here.

florisie commented:
Thank you for your comments.

I agree that your suggestion is an even nicer/cleaner solution. In the last commit I've implemented the suggested solution. However, I need to learn how to write tests in elisp (I'm quite new to elisp and not very comfortable with it) and I still need to spend a bit of time on updating the documentation. I will have a look at it soon....

florisie commented:
So I've updated the code. I changed the function:
```
(defun kubernetes-overview-set-sections (sections)
  (interactive xxxxx)

  (kubernetes-state-update-overview-sections sections)
  (kubernetes-state-trigger-redraw))
```
to
```
(defun kubernetes-overview-set-sections (sections)
  (interactive yyyy)

  (kubernetes-state-update-overview-sections sections)
  (kubernetes-state-trigger-redraw))
```
So only the "interactive part" of the function has been changed. In the  "old" interactive part there was a "completing read" with which you could read one of the options from the list of predefined views:
```
  '((overview . (context overview))
    (configmaps . (context configmaps))
    (deployments . (context deployments))
    (statefulsets . (context statefulsets))
    (ingress . (context ingress))
    (jobs . (context jobs))
    (pods . (context pods))
    (secrets . (context secrets))
    (services . (context services))
    (nodes . (context nodes))
    (persistentvolumeclaims . (context persistentvolumeclaims)))
```

I've changed it so that now you can select multiple options from the list (with a completing-read-multiple). However, I combine the result to a list of sections that always starts with a single "context" and then appends the combined cdr's of all sleceted views:

For example:
If the user selects the views pods and secrets then I combine the two list:
   (context pods)
   (context secrets)

Into a single list:
   (context pods secrets)

This all happens in the "interactive" part of the function. Therefore I do not really know how to write tests for it. I guess that if tests should be added that the "interactive part" of the function should be moved to a separate function.

jinnovation commented on kubernetes-overview.el:
```diff
@@ -374,24 +374,27 @@
-          (choice (intern (completing-read "Overview view: " names nil t))))
-     (list (alist-get choice views))))
+          (choices (completing-read-multiple "Overview view: " names nil t))
+          (list (alist-get (intern (car choices)) views)))
+     (dolist (choice (cdr choices))
+       (setq list  (append list (cdr (alist-get (intern choice) views)))))
+     (delete-dups list)))
```
Given that `list` is a reserved keyword in Elisp let's not name the variable `list` here.

Regarding your testability point, the dedupe logic here I can see us moving out into a function like `kubernetes-overview-views--dedupe` that we then call here. That would make the dedupe logic testable. I recommend this approach.

jinnovation commented on docs/how-to.md:
```diff
@@ -60,6 +60,15 @@ background processes, e.g. those for querying resources.
+|:-----------|:---------------------------------------|
+| `v`        | `M-x kubernetes-overview-select-views` |
+
+!!! info
+
+    Through `kubernetes-overview-select-views` (`v`) a list of views can be selected. The selected views will be rendered above one another in the overview buffer. Additional (custom) views can be defined through the customizable `kubernetes-overview-custom-views-alist` variable.
```
If you could provide screenshots here I think that'd make this section super helpful.

```suggestion
Through `kubernetes-overview-select-views` (`v`) a list of views can be selected. The selected views will be rendered above one another in the overview buffer.
```

Let's move the information about defining custom views into a new `Interface Customization` section elsewhere in this page.

florisie commented:
OK, changed the name of "list", moved the "dedupe" logic to a seperate function, added two (basic) tests for the dedupe function, moved the instructions to a new section on the how-to-page and added an animated gif.

noorul commented on docs/how-to.md:
```diff
@@ -31,6 +31,18 @@ comprehensive) overview of all that is possible with the package.
+
+Through `kubernetes-overview-select-views` (`v`) a list of views can be selected. The selected views will be rendered above one another in the overview buffer. Additional (custom) views can be defined through the customizable `kubernetes-overview-custom-views-alist` variable.
+
+![Selecting views](./img/views-select.gif)
```
I think you missed to add this file?

florisie replied:
Ah yes, you're right. Sorry for that. I've added it now.
